mopidy/files/library.py

```diff
@@ -0,0 +1,140 @@
+            child_path = os.path.join(local_path, dir_entry)
+            uri = path.path_to_uri(child_path)
+            printable_path = child_path.decode(sys.getfilesystemencoding(),
+                                               'ignore')
+
+            if os.path.islink(child_path) and not self._follow_symlinks:
+                logger.debug('Ignoring symlink: %s', printable_path)
+                continue
+
+            if not self._is_in_basedir(os.path.realpath(child_path)):
+                logger.debug('Ignoring symlink to outside base dir: %s',
+                             printable_path)
+                continue
+
+            if not self._show_dotfiles and dir_entry.startswith(b'.'):
+                continue
+
+            dir_entry = dir_entry.decode(sys.getfilesystemencoding(),
+                                         'replace')
+            if os.path.isdir(child_path):
+                result.append(models.Ref.directory(name=dir_entry, uri=uri))
+            elif os.path.isfile(child_path):
+                if self._is_audiofile(uri):
+                    result.append(models.Ref.track(name=dir_entry, uri=uri))
+                else:
+                    logger.debug('Ignoring non-audiofile: %s', printable_path)
+
+        result.sort(key=operator.attrgetter('name'))
+        return result
+
+    def lookup(self, uri):
+        logger.debug('looking up uri = %s', uri)
+        local_path = path.uri_to_path(uri)
+        if not self._is_in_basedir(local_path):
+            logger.warn('Ignoring URI outside base dir: %s', local_path)
+            return []
+        try:
+            result = self._scanner.scan(uri)
+            track = utils.convert_tags_to_track(result.tags).copy(
+                uri=uri, length=result.duration)
+        except exceptions.ScannerError as e:
+            logger.warning('Problem looking up %s: %s', uri, e)
+            track = models.Track(uri=uri)
+        if not track.name:
+            filename = os.path.basename(local_path)
+            name = urllib2.unquote(filename).decode(
+                sys.getfilesystemencoding(), 'ignore')
+            track = track.copy(name=name)
+        return [track]
+
+    def _get_media_dirs(self, config):
+        for entry in config['files']['media_dirs']:
+            media_dir = {}
+            media_dir_split = entry.split('|', 1)
+            local_path = path.expand_path(
+                media_dir_split[0].encode(sys.getfilesystemencoding()))
+            if not local_path:
+                logger.warn('Could not expand path %s', media_dir_split[0])
+                continue
+            elif not os.path.isdir(local_path):
+                logger.warn('%s is not a directory', local_path)
+                continue
+            media_dir['path'] = local_path
+            if len(media_dir_split) == 2:
+                media_dir['name'] = media_dir_split[1]
+            else:
+                media_dir['name'] = media_dir_split[0].replace(os.sep, '+')
```

Comment:
I suspect the `replace()` here is because MPD don't like `/` in dir names? If so, this is something we should rewrite in the MPD frontend instead. If I'm right, please add a TODO comment here explaining why this is here and where it should be fixed instead. I can live with doing the replace here for now.